lattices/src/semiring_application.rs

```diff
@@ -0,0 +1,537 @@
+    /// Infinity
+    Infinity,
+    /// Natural numbers
+    Finite(u32),
+}
+
+#[allow(dead_code)]
+#[derive(Clone, Copy, PartialEq, Debug)]
+/// Implementation of the Cost/Tropical semiring (N U Inf, min, +, inf, 0)
+pub struct Cost(U32WithInfinity);
+
+impl Cost {
+    /// Create a new instance of Cost.
+    pub fn new(value: U32WithInfinity) -> Self {
+        Cost(value)
+    }
+}
+
+/// Implementation of the addition trait for Cost semiring.
+impl Addition<Cost> for Cost {
+    /// Min operation
+    fn add(&mut self, other: Self) {
+        match (self.0, other.0) {
+            (U32WithInfinity::Infinity, _) => self.0 = other.0,
+            (_, U32WithInfinity::Infinity) => (),
+            (U32WithInfinity::Finite(a), U32WithInfinity::Finite(b)) => {
+                if a < b {
+                    self.0 = U32WithInfinity::Finite(a);
+                } else {
+                    self.0 = U32WithInfinity::Finite(b);
+                }
+            }
+        }
+    }
+}
+
+/// Implementation of the multiplication trait for Cost semiring.
+impl Multiplication<Cost> for Cost {
+    /// + operation
+    fn mul(&mut self, other: Cost) {
+        match (self.0, other.0) {
+            (U32WithInfinity::Infinity, _) | (_, U32WithInfinity::Infinity) => {
+                self.0 = U32WithInfinity::Infinity;
+            }
+            (U32WithInfinity::Finite(a), U32WithInfinity::Finite(b)) => {
+                self.0 = U32WithInfinity::Finite(a + b);
+            }
+        }
+    }
+}
+
+/// Implementation of Identity for addition.
+impl Zero<U32WithInfinity> for Cost {
+    /// Infinity is the identity element for addition operation.
+    fn zero(&self) -> Cost {
+        Cost(U32WithInfinity::Infinity)
+    }
+}
+
+/// Implementation of Identity for multiplication.
+impl One<U32WithInfinity> for Cost {
+    /// 0 is the one element of the semiring.
+    fn one(&self) -> Cost {
+        Cost(U32WithInfinity::Finite(0))
+    }
+}
+
+#[allow(dead_code)]
+/// use Tropical Semiring implementation for calculating All Pair Shortest Path in a graph with positive edge weights.
+/// P(x,y) = shortest path from x to y
+/// E(x,y) = edge weight from x to y
+/// P(x,y) = min(E(x,y), min z (P(x,z) + E(z,y)) where (min,+) are the “addition” and “multiplication” operators in the tropical semirings
+fn all_pairs_shortest_path(graph: Vec<Vec<U32WithInfinity>>) -> Vec<Vec<u32>> {
+    let n = graph.len();
+    let mut dist: Vec<Vec<Cost>> = vec![vec![Cost::new(U32WithInfinity::Infinity); n]; n];
+    for i in 0..n {
+        for j in 0..n {
+            dist[i][j] = Cost::new(graph[i][j]);
+        }
+    }
+    for i in 0..n {
+        for j in 0..n {
+            for k in 0..n {
+                // distance[i][j] = min(distance[i][j], distance[i][k] + distance[k][j])
+                let mut new_cost = dist[i][k];
+                new_cost.mul(dist[k][j]);
+                dist[i][j].add(new_cost);
+            }
+        }
+    }
+
```

Comment:
Done.

Reply:
Why is it that we should remove infinity distances?

Reply:
infinity is an encoding of there not existing any path between the two vertices.  We don't need the infinities in the output because its kind of an internal implementation detail that we encode it that way during the computation. So its just a preference for what the output looks like to remove them so the output is actual paths.